File: assets/scripts/cfdcalculator.js
const reoutput = document.querySelector('.reoutput');
const yhoutput = document.querySelector('.yhoutput');
const bloutput = document.querySelector('.bloutput');
const noutput = document.querySelector('.noutput');
const floutput = document.querySelector('.floutput');
const mnoutput = document.querySelector('.mnoutput');
const tioutput = document.querySelector('.tioutput');
const lmoutput = document.querySelector('.lmoutput');
const tkeoutput = document.querySelector('.tkeoutput');
const tdoutput = document.querySelector('.tdoutput');
const nutoutput = document.querySelector('.nutoutput');
const sdroutput = document.querySelector('.sdroutput');
document.getElementById("cfdCalculatorForm").addEventListener('submit', cfdCalculatorFunction, false);
function cfdCalculatorFunction(event) {
    var gr = document.getElementById("gr").value;
    var density = document.getElementById("density").value;
    var viscosity = document.getElementById("viscosity").value;
    var velocity = document.getElementById("velocity").value;
    var length = document.getElementById("length").value;
    var yplus = document.getElementById("yplus").value;
    var re = ((density * velocity * length) / viscosity);
    var cf = Math.pow(((2 * Math.log10(re)) - 0.65), -2.3);
    var tw = (1 / 2) * density * Math.pow(velocity, 2) * cf;
    var ut = Math.sqrt(tw / density);
    var yp = (yplus * viscosity) / (ut * density);
    var yh = 2 * yp;
    reoutput.textContent = 'Reynolds Number=' + re.toFixed(8);
    yhoutput.textContent = 'First Mesh Layer Height=' + yh.toFixed(8);
    if (re <= 500000) {
        var bl = ((4.91 * length) / Math.sqrt(re));
    }
    else {
        var bl = ((0.38 * length) / Math.pow(re, (1 / 5)));
    }
    bloutput.textContent = 'Boundary Layer Height=' + bl.toFixed(8);
    if ((gr >= 1.01) && (gr <= 2)) {
        var k = yh;
        for (n = 0; n <= 99; n++) {
            k = gr * k;
            if (k >= bl) {
                break;
            }
        }
        var fl = k - (k / gr);
        var mn = gr * k;
        noutput.textContent = 'Number of Layers=' + n;
        floutput.textContent = 'Final Layer Thickness=' + fl.toFixed(8);
        mnoutput.textContent = 'Minimum Mesh Size=' + mn.toFixed(8);
    }
    else {
        noutput.textContent = 'Wrong Growth Ratio';
    }
    var cmu = 0.09;
    var ti = 0.055 * Math.pow(re, -0.041);
    var lm = 0.07 * length;
    var tke = (3 / 2) * Math.pow((velocity * ti), 2);
    var td = ((Math.pow(cmu, (3 / 4))) * (Math.pow(tke, (3 / 2)))) / lm;
    var nut = (cmu * (Math.pow(tke, 2))) / td;
    var sdr = ((Math.pow(cmu, (-1 / 4))) * (Math.pow(tke, (1 / 2)))) / lm;
    tioutput.textContent = 'Turbulence Intensity=' + ti.toFixed(8);
    lmoutput.textContent = 'Length Scale/Mixing Length=' + lm.toFixed(8);
    tkeoutput.textContent = 'Turbulent Kinetic Energy=' + tke.toFixed(8);
    tdoutput.textContent = 'Turbulent Dissipation=' + td.toFixed(8);
    nutoutput.textContent = 'Turbulent Viscosity=' + nut.toFixed(8);
    sdroutput.textContent = 'Specific Dissipation Rate=' + sdr.toFixed(8);
    event.preventDefault();
    return 0;
}

File: theme/calculators/pages/cfd-calculator.py
import streamlit as st
import numpy as np

def cfd_calculator():
    """
    This function performs CFD calculations based on user input.
    """
    with st.sidebar:
        st.title("CFD Calculator")
        # Get user input
        st.subheader("Enter the following parameters:")
        gr = st.number_input("Growth Ratio (GR)", min_value=1.0, max_value=2.0, step=0.01)
        density = st.number_input("Density (rho)", min_value=0.1)
        viscosity = st.number_input("Viscosity (mu)", min_value=0.0000001)
        velocity = st.number_input("Velocity (u)", min_value=0.1)
        length = st.number_input("Length (L)", min_value=0.00001)
        yplus = st.number_input("Y Plus (y+)", min_value=1)
        # Perform calculations
        re = density * velocity * length / viscosity
        cf = (2 * np.log10(re) - 0.65)**(-2.3)
        tw = 0.5 * density * velocity**2 * cf
        ut = np.sqrt(tw / density)
        yp = (yplus * viscosity) / (ut * density)
        yh = 2 * yp
        # Boundary Layer Height calculation
        if re <= 500000:
            bl = 4.91 * length / np.sqrt(re)
        else:
            bl = 0.38 * length / re**(1/5)

        # Check for valid growth ratio and perform calculations if applicable
        if 1.0 <= gr <= 2.0:
            k = yh
            for _ in range(100):  # Limit iterations for efficiency
                k *= gr
                if k >= bl:
                    break
            fl = k - k / gr
            mn = gr * k
            n = _ + 1  # Number of iterations is the number of layers
        else:
            n = None
            fl = None
            mn = None
            st.error("Wrong Growth Ratio")

        # Other calculations (assuming constant values for simplicity)
        cmu = 0.09
        ti = 0.055 * re**(-0.041)
        lm = 0.07 * length
        tke = (3 / 2) * ((velocity * ti)**2)
        td = ((cmu**(3 / 4)) * (tke**(3 / 2))) / lm
        nut = (cmu * (tke**2)) / td
        sdr = ((cmu**(-1 / 4)) * (tke**(1 / 2))) / lm
  
    # Display results
    st.subheader("Results:")
    st.metric("Reynolds Number (Re)", f"{re:.8f}")
    st.metric("First Mesh Layer Height (yh)", f"{yh:.8f}")
    st.metric("Boundary Layer Height (bl)", f"{bl:.8f}")
    if n is not None:
        st.metric("Number of Layers (n)", n)
        st.metric("Final Layer Thickness (fl)", f"{fl:.8f}")
        st.metric("Minimum Mesh Size (mn)", f"{mn:.8f}")
    st.metric("Turbulence Intensity (TI)", f"{ti:.8f}")
    st.metric("Length Scale/Mixing Length (lm)", f"{lm:.8f}")
    st.metric("Turbulent Kinetic Energy (TKE)", f"{tke:.8f}")
    st.metric("Turbulent Dissipation (td)", f"{td:.8f}")
    st.metric("Turbulent Viscosity (nut)", f"{nut:.8f}")
    st.metric("Specific Dissipation Rate (SDR)", f"{sdr:.8f}")

# Main app
st.set_page_config(layout="wide")
cfd_calculator()
st.text('This code/calculator/service/product is subject to the terms of the MIT License')